src/client.js

```diff
@@ -36,17 +36,60 @@ function initSocket() {
 
 global.socket = initSocket();
 
-function createElement(Component, props) {
-  if (Component.fetchData) {
-    Component.fetchData(store.getState, store.dispatch,
-                        props.location, props.params);
+const history = useScroll(createHistory)({routes: createRoutes(routes)});
+syncReduxAndRouter(history, store);
+
+let lastMatchedLocBefore;
+let lastMatchedLocAfter;
+
+history.listenBefore((location, callback) => {
```

Comment:
I don't think you should be using `match` browser-side at all. We should be using a `history` that has been augmented with `useRoutes`: https://github.com/rackt/react-router/blob/master/docs/API.md#useroutescreatehistory

In fact you pass `routes` into `createHistory` but that's not doing anything. You need to also wrap `createHistory` with `useRoutes`, and then the `listen` signature changes: you get `(error, nextState) => {}` instead, and you can just use `nextState` to get all the components.